test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -510,6 +510,47 @@ TEST_VM(os, release_multi_mappings) {
 }
 #endif // !AIX
 
+#ifndef _AIX // JDK-8257041
+TEST_VM(os, partial_release_multi_mappings) {
+
+  //  see "release_multi_mappings" test comment for general explanation
+
+  //  thread grabbing that memory.
+
+  const size_t stripe_len = os::vm_allocation_granularity();
+  const int num_stripes = 4;
+  const size_t total_range_len = stripe_len * num_stripes;
+  const size_t partial_release_len = total_range_len - (stripe_len/2); // partially release the last stripe
+
+  // reserve address space...
+  address p = reserve_multiple(num_stripes, stripe_len);
+  ASSERT_NE(p, (address)NULL);
+  PRINT_MAPPINGS("A");
+
+  // .. release the middle stripe...
+  address p_middle_stripes = p + stripe_len;
+  const size_t middle_stripe_len = (num_stripes - 2) * stripe_len;
+  {
+    // On Windows, temporarily switch on UseNUMAInterleaving to allow release_memory to release
+    //  multiple mappings in one go (otherwise we assert, which we test too, see death test below).
+    WINDOWS_ONLY(NUMASwitcher b(true);)
+    ASSERT_TRUE(os::release_memory((char*)p_middle_stripes, middle_stripe_len));
+  }
+  PRINT_MAPPINGS("B");
+
+  // ...re-reserve the middle stripes. This should work unless release silently failed.
+  address p2 = (address)os::attempt_reserve_memory_at((char*)p_middle_stripes, middle_stripe_len);
+  ASSERT_EQ(p2, p_middle_stripes);
+  PRINT_MAPPINGS("C");
+
+  // Clean up. Release all mappings except a 1/2 stripe sized portion at the very end.
+  {
+    WINDOWS_ONLY(NUMASwitcher b(true);) // allow release_memory to release multiple regions
+    ASSERT_TRUE(os::release_memory((char*)p, partial_release_len));
+  }
+}
+#endif // !AIX
+
```

Comment:
- This tests case B, and we need case C and D too.
- You should make the first - middle - release the testing release with the partwise releasing. The final release is just to clean up the mess. For that to work, the stripes must be larger and a multiple of os::allocation_granularity, since otherwise you cannot release them partwise.

Would be cool if you fit all of these cases into a single function, e.g. 
```
static void test_multi_release(size_t stripe_size, int num_stripes, size_t release_range_offset, size_t release_range_size)
```

And then call them 4 times in 4 tests, for each of the test cases. E.g. with a stripe len set to 1M could be like this:

```
TEST_VM(TEST_VM, release_multi_mappings_full_regions) {
  test_multi_release(M, 4, M, 2*M); // release stripes 2 and 3 fully
}

TEST_VM(TEST_VM, release_multi_mappings_partial_regions) {
  test_multi_release(M, 4, M + 512*K, 512*K + M); // release stripe 2 partly, stripe 3 fully
}

TEST_VM(TEST_VM, release_multi_mappings_partial_regions) {
  test_multi_release(M, 4, M, M + 512*K); // release stripe 2 fully, stripe 3 partly
}

TEST_VM(TEST_VM, release_multi_mappings_partial_regions) {
  test_multi_release(M, 4, M + 256K, 512 * K); // release a 512K hole into stripe 2
}

```

Reply:
I tried to make a picture of the memory and the parts that should get released for testing:

```
//  allocate              [aaaa][bbbb][cccc][dddd]

//  1 release partial     [....][bbbb][cccc][....]
//  2 release partial     [....][..bb][cccc][....]
//  3 release partial     [....][bb..][cccc][....]
//  4 release partial     [....][bbbb][cc..][....]
//  5 release partial     [....][bbbb][..cc][....]
//  6 release partial     [....][..bb][cc..][....]
//  7 release partial     [....][bb..][..cc][....]
//  8 release partial     [....][bb..][cc..][....]
//  9 release partial     [....][..bb][..cc][....]
// 10 release partial     [....][.bb.][....][....]
// 11 release partial     [....][.bb.][.cc.][....]
```

Are we still missing any cases?

Did I add too much? (i.e. drop case 11?)

Reply:
I think I see 3 more cases for the release pattern that we can add:

```
  [aaaa][bbbb][cccc][dddd]

1 [....][xxxx][xxxx][....]
2 [....][..xx][xxxx][....]
3 [....][xxxx][xx..][....]
4 [....][..xx][xx..][....]

5 [....][..xx][....][....]
6 [....][....][xx..][....]
7 [....][.xx.][....][....]
```

Aren't cases 5,6 and 7 unique?

Reply:
> Aren't cases 5,6 and 7 unique?

Are they not already covered by existing tests? Since that functionality already exists?

Reply:
I might be missing it, but `grep -crn "release_memory(" test/hotspot/gtest` shows:

```
test/hotspot/gtest/memory/test_virtualspace.cpp:2
test/hotspot/gtest/runtime/test_os.cpp:20
test/hotspot/gtest/runtime/test_committed_virtualmemory.cpp:1
test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp:1
```

And when I look at the files I see:

`test/hotspot/gtest/runtime/test_os.cpp`
- tests bad ranges on Windows
- tests striped reserve, followed by one full release, one full re-reserve, then release again
- tests os::print_memory_mappings()
- tests os::win32::find_mapping()
- tests `stripe-stripe-stripe-stripe-stripe-stripe` reserve, followed by `stripe-______-______-______-______-stripe` release, then re-reserve middle stripes, then full release

`test/hotspot/gtest/memory/test_virtualspace.cpp`
- tests different sizes, alignments, one at a time

`test/hotspot/gtest/runtime/test_committed_virtualmemory.cpp`
- tests whether the memory is resident

`test/hotspot/gtest/nmt/test_nmt_locationprinting.cpp`
- tests NMT print_containing_region mechanism

So, no, I don't see dedicated partial release tests of single ranges.

Reply:
You may be right, then. If you feel up for it, cases 5-7 would also be good to test. With the proposed setup, it would not be much more work. Thanks!